### DISCUSSION TITLE
C API expose column_types for duckdb_appender

### DISCUSSION THREAD

**Discussion by karlseguin** [duckdb/duckdb#8069] [Ideas]
There's currently no way to get the `duckdb_types` for an appender, which makes binding dynamic data ambiguous.  For example, consider you're given this JSON array that you want to append:

```json
[
  [1, "b3ZlcjkwMDA="]
]
```

If you use a prepared statement, you can use `duckdb_param_type` to see that the 2nd parameter is a `DUCKDB_TYPE_BLOB` and base64 decode the value. There are other common ambiguities that a generic implementation (like a driver), might face, such as timestamp vs i64.

It would be helpful to copy the `duckdb_nparams` and `duckdb_param_type` from the `duckdb_prepared_statement` to the `duckdb_appender`.

**Comment by karlseguin** [comment]
https://github.com/duckdb/duckdb/issues/3923 is a real-life example where the ambiguity has obviously caused some issues.

**Comment by ak-coram** [comment]
I'm currently using this workaround in the CL bindings (creating an insert statement for the table to determine the types):

https://github.com/ak-coram/cl-duckdb/blob/14c136c77897aa49f84c7982d19b91de1579e9bc/duckdb.lisp#L718C13-L718C13

**Comment by Giorgi** [comment]
There is now https://duckdb.org/docs/api/c/appender#duckdb_appender_column_type
